File: src/util/file_utils.c
/*
 * File Utilities
 * Helper functions for file I/O operations
 */

#ifndef __APPLE__
    #define _GNU_SOURCE
#endif

#include "file_utils.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <bsd/string.h>
#include <sys/stat.h>
#include <errno.h>
#include <limits.h>
#include <libgen.h>

#ifndef TEST_BUILD
#include "../logger.h"
#else
// Test build stubs
#define LOG_ERROR(fmt, ...) ((void)0)
#define LOG_DEBUG(fmt, ...) ((void)0)
#endif

/**
 * Read entire file into memory
 */
char* read_file(const char *path) {
    FILE *f = fopen(path, "rb");
    if (!f) return NULL;

    fseek(f, 0, SEEK_END);
    long fsize = ftell(f);
    if (fsize < 0) {
        fclose(f);
        return NULL;
    }
    fseek(f, 0, SEEK_SET);

    char *content = malloc((size_t)fsize + 1);
    if (content) {
        size_t bytes_read = fread(content, 1, (size_t)fsize, f);
        (void)bytes_read; // Suppress unused result warning
        content[fsize] = 0;
    }

    fclose(f);
    return content;
}

/**
 * Create directory recursively (like mkdir -p)
 */
int mkdir_p(const char *path) {
    char tmp[PATH_MAX];
    char *p = NULL;
    size_t len;

    if (strlcpy(tmp, path, sizeof(tmp)) >= sizeof(tmp)) {
        return -1; // Path too long
    }
    len = strlen(tmp);

    // Remove trailing slash
    if (len > 0 && tmp[len - 1] == '/') {
        tmp[len - 1] = '\0';
    }

    // Create directories recursively
    for (p = tmp + 1; *p; p++) {
        if (*p == '/') {
            *p = '\0';
            if (mkdir(tmp, 0755) != 0 && errno != EEXIST) {
                return -1;
            }
            *p = '/';
        }
    }

    // Create final directory
    if (mkdir(tmp, 0755) != 0 && errno != EEXIST) {
        return -1;
    }

    return 0;
}

/**
 * Write content to file
 */
int write_file(const char *path, const char *content) {
    // Create parent directories if they don't exist
    char *path_copy = strdup(path);
    if (!path_copy) return -1;

    // Extract directory path
    char *dir_path = dirname(path_copy);

    // Create directory recursively (ignore errors if directory already exists)
    (void)mkdir_p(dir_path); // Ignore errors - directory may already exist

    free(path_copy);

    // Now try to open/create the file
    FILE *f = fopen(path, "wb");
    if (!f) return -1;

    size_t len = strlen(content);
    size_t written = fwrite(content, 1, len, f);
    fclose(f);

    return (written == len) ? 0 : -1;
}

/**
 * Resolve path (handle relative paths and canonicalization)
 */
char* resolve_path(const char *path, const char *working_dir) {
    // Join with working_dir if relative; attempt to canonicalize if possible.
    char joined[PATH_MAX];
    if (path[0] == '/') {
        snprintf(joined, sizeof(joined), "%s", path);
    } else {
        snprintf(joined, sizeof(joined), "%s/%s", working_dir, path);
    }

    // Try to canonicalize. This succeeds only if the path (or its parents) exist.
    char *clean = realpath(joined, NULL);
    if (clean) {
        return clean; // Caller takes ownership
    }

    // Fall back to the joined path even if parent dirs don't exist.
    // This enables tools like Write to create missing directories (mkdir -p in write_file).
    return strdup(joined);
}

/**
 * Save binary data to file
 */
int save_binary_file(const char *filename, const void *data, size_t size) {
    if (!filename || !data || size == 0) {
        return -1;
    }

    // Create parent directories if they don't exist
    char *path_copy = strdup(filename);
    if (!path_copy) return -1;

    // Extract directory path
    char *dir_path = dirname(path_copy);

    // Create directory recursively (ignore errors if directory already exists)
    (void)mkdir_p(dir_path); // Ignore errors - directory may already exist

    free(path_copy);

    // Open file for writing
    FILE *f = fopen(filename, "wb");
    if (!f) {
        LOG_ERROR("Failed to open file for binary writing: %s", filename);
        return -1;
    }

    // Write binary data
    size_t written = fwrite(data, 1, size, f);
    fclose(f);

    if (written != size) {
        LOG_ERROR("Failed to write all binary data to file: %s (written: %zu, expected: %zu)",
                 filename, written, size);
        return -1;
    }

    LOG_DEBUG("Successfully saved binary data to '%s' (%zu bytes)", filename, size);
    return 0;
}

File: zig/util/file_utils.zig
//! File Utilities
//!
//! Idiomatic Zig replacements for src/util/file_utils.c
//!
//! Key C→Zig translations:
//!   - `read_file`          → `readFile`      (readToEndAlloc, proper error union)
//!   - `write_file`         → `writeFile`     (std.fs, creates dirs automatically)
//!   - `mkdir_p`            → `mkdirP`        (std.fs.Dir.makePath)
//!   - `resolve_path`       → `resolvePath`   (std.fs.path.resolve)
//!   - `save_binary_file`   → `saveBinaryFile`(writeFile variant for binary slices)
//!
//! All functions that return owned memory accept an allocator parameter.
//! The C version used global state and strdup — here ownership is explicit.

const std = @import("std");

/// Maximum file size we'll read into memory (64 MiB).
/// This prevents accidentally exhausting RAM on huge files.
pub const max_read_size: usize = 64 * 1024 * 1024;

// ---------------------------------------------------------------------------
// Core operations
// ---------------------------------------------------------------------------

/// Read the entire contents of `path` into an allocated slice.
/// Caller must free with `allocator.free`.
pub fn readFile(allocator: std.mem.Allocator, path: []const u8) ![]u8 {
    const file = try std.fs.openFileAbsolute(path, .{});
    defer file.close();
    return file.readToEndAlloc(allocator, max_read_size);
}

/// Read the entire contents of a path relative to a directory.
pub fn readFileRelative(
    allocator: std.mem.Allocator,
    dir: std.fs.Dir,
    relative_path: []const u8,
) ![]u8 {
    const file = try dir.openFile(relative_path, .{});
    defer file.close();
    return file.readToEndAlloc(allocator, max_read_size);
}

/// Write `content` to `path`, creating parent directories as needed.
/// Overwrites existing files.
pub fn writeFile(path: []const u8, content: []const u8) !void {
    // Ensure parent directories exist
    if (std.fs.path.dirname(path)) |dir_path| {
        try mkdirP(dir_path);
    }
    const file = try std.fs.createFileAbsolute(path, .{ .truncate = true });
    defer file.close();
    try file.writeAll(content);
}

/// Write binary `data` to `path`, creating parent directories as needed.
pub fn saveBinaryFile(path: []const u8, data: []const u8) !void {
    return writeFile(path, data);
}

/// Create a directory and all its parents (like `mkdir -p`).
/// Succeeds silently if the directory already exists.
pub fn mkdirP(path: []const u8) !void {
    // std.fs.makeDirAbsolute errors on EEXIST — use cwd().makePath for
    // recursive creation, which handles both absolute and relative paths.
    std.fs.cwd().makePath(path) catch |err| switch (err) {
        error.PathAlreadyExists => {}, // fine
        else => return err,
    };
}

// ---------------------------------------------------------------------------
// Path helpers
// ---------------------------------------------------------------------------

/// Resolve `path` against `working_dir` when it is relative, then
/// canonicalize via `realpath` when the path exists.
///
/// When canonicalization fails (e.g. the target doesn't exist yet), returns a
/// simple joined path — this enables tools like Write to create new files.
///
/// Caller must free the result with `allocator.free`.
pub fn resolvePath(
    allocator: std.mem.Allocator,
    path: []const u8,
    working_dir: []const u8,
) ![]u8 {
    // If absolute, try realpath then fall back to a copy.
    if (std.fs.path.isAbsolute(path)) {
        return std.fs.realpathAlloc(allocator, path) catch
            allocator.dupe(u8, path);
    }
    // Join working_dir + path, then try realpath.
    const joined = try std.fs.path.join(allocator, &.{ working_dir, path });
    defer allocator.free(joined);

    return std.fs.realpathAlloc(allocator, joined) catch
        allocator.dupe(u8, joined);
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

test "mkdirP: creates nested directories" {
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    const sub = "a/b/c";
    try tmp.dir.makePath(sub);
    // Verify the leaf exists
    var leaf = try tmp.dir.openDir(sub, .{});
    defer leaf.close();
}

test "writeFile and readFile: round-trip" {
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    const allocator = std.testing.allocator;

    // Build an absolute path inside the temp dir
    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const tmp_path = try tmp.dir.realpath(".", &path_buf);
    const file_path = try std.fs.path.join(allocator, &.{ tmp_path, "test.txt" });
    defer allocator.free(file_path);

    const content = "hello, zig!\n";
    try writeFile(file_path, content);

    const read_back = try readFile(allocator, file_path);
    defer allocator.free(read_back);

    try std.testing.expectEqualStrings(content, read_back);
}

test "writeFile: creates parent directories automatically" {
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    const allocator = std.testing.allocator;

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const tmp_path = try tmp.dir.realpath(".", &path_buf);
    const file_path = try std.fs.path.join(allocator, &.{ tmp_path, "nested/dir/file.txt" });
    defer allocator.free(file_path);

    try writeFile(file_path, "content");

    const read_back = try readFile(allocator, file_path);
    defer allocator.free(read_back);
    try std.testing.expectEqualStrings("content", read_back);
}

test "resolvePath: absolute path returned as-is (when exists)" {
    const allocator = std.testing.allocator;
    // /tmp always exists
    const resolved = try resolvePath(allocator, "/tmp", "/ignored");
    defer allocator.free(resolved);
    // realpath may resolve /tmp → /private/tmp on macOS, so just check it starts with /
    try std.testing.expect(resolved[0] == '/');
}

test "resolvePath: relative path joined with working_dir" {
    const allocator = std.testing.allocator;
    // Use a path that doesn't exist — falls back to the joined string
    const resolved = try resolvePath(allocator, "subdir/file.txt", "/some/base");
    defer allocator.free(resolved);
    // Should contain "subdir/file.txt" somewhere
    try std.testing.expect(std.mem.indexOf(u8, resolved, "subdir/file.txt") != null);
}
